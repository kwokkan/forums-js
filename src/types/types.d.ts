
interface IForumSessionUser {
    id: number;
    name: string;
    joinedDate: number;
}

interface INextSession extends any {
    forumsUser?: IForumSessionUser;
}

interface INextProvider extends any {
    id: string;
    name: string;
}

declare module "next-auth" {
    interface INextAccount extends any {
        provider: string;
        id: string;
    }

    export interface IProviderOptions {
        providers: any[];
        callbacks?: IProviderOptionsCallback;
        events?: INextEventsOptions;
        pages?: INextPageOptions;
        debug?: boolean;
    }

    interface IProviderOptionsCallback {
        signIn?: (user: INextUser, account: INextAccount, profile: INextProfile) => Promise<bool>;
        redirect?: (url: string, baseUrl: string) => Promise<string>;
        session?: (session: INextSession, user: INextUser, sessionToken: string) => Promise<INextSession>;
        jwt?: (token: INextToken, user: INextUser, account: INextAccount, profile: INextProfile, isNewUser: boolean) => Promise<INextToken>;
    }

    interface INextEventsOptions {
        signIn?: (message: any) => Promise;
        signOut?: (message: any) => Promise;
        createUser?: (message: any) => Promise;
        linkAccount?: (message: any) => Promise;
        session?: (message: any) => Promise;
        error?: (message: any) => Promise;
    }

    interface INextPageOptions {
        signIn?: string;
        signOut?: string;
        error?: string;
        verifyRequest?: string;
        newUser?: string;
    }

    interface INextProfile extends any {
    }


    interface INextToken extends any {
        forumsUser?: IForumSessionUser;
    }

    interface INextUser extends any {
        name: string;
        forumsUser?: IForumSessionUser;
    }

    function handler(req: IncomingMessage, res: ServerResponse, options: IProviderOptions): void;

    export = handler;
}

declare module "next-auth/client" {
    export const Provider: any;

    export function getSession(context: { req: IncomingMessage }): Promise<INextSession>;

    export function providers<T>(context: GetServerSidePropsContext<T>): Promise<INextProvider[]>;

    export function signIn(provider?: string, args?: { callbackUrl?: string }): void;

    export function signOut(): void;

    export function useSession(): [INextSession | null, boolean];
}

declare module "next-auth/providers" {
    interface IProviderConfig {
        clientId: string,
        clientSecret?: string,
        scope?: string[] | string
    }

    export = {
        GitHub: (config: IProviderConfig) => any
    };
}
