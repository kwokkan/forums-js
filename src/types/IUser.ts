export interface IUser {
    id: number;
    name: string;
    joinedDate: Date;
}
