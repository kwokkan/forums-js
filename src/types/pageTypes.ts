import { GetServerSideProps } from "next";
import { ParsedUrlQuery } from "querystring";

export type GetTypedServerSideProps<TParams extends ParsedUrlQuery, TProps = { [key: string]: any }> = GetServerSideProps<TProps, TParams>;
