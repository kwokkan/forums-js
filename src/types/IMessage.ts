import { IUser } from "./IUser";

export interface IMessage {
    id: number;
    created: Date;
    content: string;
    user: IUser;
}
