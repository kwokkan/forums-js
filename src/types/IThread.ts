import { IMessage } from "./IMessage";

export interface IThread {
    id: number;
    created: number;
    name: string;
    messages: IMessage[];
}
