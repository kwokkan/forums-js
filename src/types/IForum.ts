import { IThread } from "./IThread";

export interface IForum {
    id: number;
    created: Date;
    name: string;
    description?: string;
    threads?: IThread[];
}
