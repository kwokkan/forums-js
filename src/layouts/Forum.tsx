import Link from "next/link";
import React, { Fragment } from "react";
import { Button, Content, Pagination, Panel, PanelGroup, Table } from "rsuite";
import { IForum } from "../types/IForum";
import { IUser } from "../types/IUser";
import { formatDate } from "../utils/dateUtils";

const { Column, HeaderCell, Cell } = Table;

export interface IProps {
    forum: IForum;
    user?: IUser;
}

export function Forum({ forum, user }: IProps) {
    return (
        <Panel header={<h3>{forum.name}</h3>}>
            <Content>
                {forum.description &&
                    <Fragment>
                        <p>
                            {forum.description}
                        </p>

                        <hr />
                    </Fragment>
                }

                {user &&
                    <PanelGroup>
                        <Link href={`/forums/${forum.id}/new`}>
                            <a>
                                <Button appearance="primary">
                                    New Thread
                                </Button>
                            </a>
                        </Link>
                    </PanelGroup>
                }

                {forum.threads &&
                    <PanelGroup>
                        <Table data={forum.threads} autoHeight wordWrap>
                            <Column flexGrow={11}>
                                <HeaderCell>Threads</HeaderCell>
                                <Cell>{(x) => <Link href={`/threads/${x.id}`}><a>{x.name}</a></Link>}</Cell>
                            </Column>

                            <Column flexGrow={1}>
                                <HeaderCell>Last post</HeaderCell>
                                <Cell>{(x) => formatDate(x.created)}</Cell>
                            </Column>
                        </Table>

                        <Pagination limit={10} total={forum.threads.length}>
                        </Pagination>
                    </PanelGroup>
                }
            </Content>
        </Panel>
    );
}
