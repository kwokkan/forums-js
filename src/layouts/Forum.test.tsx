import React from "react";
import { create } from "react-test-renderer";
import { IForum } from "../types/IForum";
import { IUser } from "../types/IUser";
import { Forum } from "./Forum";

test("Renders with nothing", () => {
    const forum: IForum = {
        id: 1,
        created: new Date(2020, 1, 1).getTime(),
        name: "First forum"
    };

    const tree = create(
        <Forum forum={forum} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

test("Renders with description", () => {
    const forum: IForum = {
        id: 1,
        created: new Date(2020, 1, 1).getTime(),
        name: "First forum",
        description: "Forum description."
    };

    const tree = create(
        <Forum forum={forum} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

test("Renders with new thread button", () => {
    const forum: IForum = {
        id: 1,
        created: new Date(2020, 1, 1).getTime(),
        name: "First forum",
        description: "Forum description."
    };

    const user: IUser = {
        id: 1,
        joinedDate: 2,
        name: "Test user"
    };

    const tree = create(
        <Forum forum={forum} user={user} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

/*
test("Renders with threads", () => {
    const forum: IForum = {
        id: 1,
        created: new Date(2020, 1, 1).getTime(),
        name: "First forum",
        threads: [
            {
                id: 2,
                created: new Date(2020, 2, 2).getTime(),
                name: "First thread",
                messages: []
            }
        ]
    };

    const tree = create(
        <Forum forum={forum} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
*/
