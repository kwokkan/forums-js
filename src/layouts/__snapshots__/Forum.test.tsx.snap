// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with description 1`] = `
<div
  className="rs-panel"
>
  <div
    aria-expanded={false}
    className="rs-panel-header"
    tabIndex={-1}
  >
    <h3
      className="rs-panel-title"
    >
      First forum
    </h3>
  </div>
  <div
    className="rs-panel-body"
    role="region"
  >
    <main
      className="rs-content"
    >
      <p>
        Forum description.
      </p>
      <hr />
    </main>
  </div>
</div>
`;

exports[`Renders with new thread button 1`] = `
<div
  className="rs-panel"
>
  <div
    aria-expanded={false}
    className="rs-panel-header"
    tabIndex={-1}
  >
    <h3
      className="rs-panel-title"
    >
      First forum
    </h3>
  </div>
  <div
    className="rs-panel-body"
    role="region"
  >
    <main
      className="rs-content"
    >
      <p>
        Forum description.
      </p>
      <hr />
      <div
        className="rs-panel-group"
      >
        <a
          href="/forums/1/new"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          <button
            className="rs-btn rs-btn-primary"
            type="button"
          >
            New Thread
            <span
              className="rs-ripple-pond"
            >
              <span
                className="rs-ripple"
              />
            </span>
          </button>
        </a>
      </div>
    </main>
  </div>
</div>
`;

exports[`Renders with nothing 1`] = `
<div
  className="rs-panel"
>
  <div
    aria-expanded={false}
    className="rs-panel-header"
    tabIndex={-1}
  >
    <h3
      className="rs-panel-title"
    >
      First forum
    </h3>
  </div>
  <div
    className="rs-panel-body"
    role="region"
  >
    <main
      className="rs-content"
    />
  </div>
</div>
`;
